File: pkg/homekit_adapter.py
"""HomeKit adapter for Mozilla IoT Gateway."""

from gateway_addon import Adapter
from hapclient import HapClient

from .homekit_database import HomeKitDatabase
from .homekit_device import (HomeKitBridge, HomeKitBulb, HomeKitPlug,
                             HomeKitUnpairedDevice)
from .homekit_errors import DatabaseError, HapError, PairingError

try:
    from gateway_addon.errors import SetPinError
except ImportError:
    pass

_TIMEOUT = 3


class HomeKitAdapter(Adapter):
    """Adapter for HomeKit devices."""

    def __init__(self, verbose=False):
        """
        Initialize the object.

        config -- current user configuration
        verbose -- whether or not to enable verbose logging
        """
        self.name = self.__class__.__name__
        Adapter.__init__(self,
                         'homekit-adapter',
                         'homekit-adapter',
                         verbose=verbose)

        self.pairing = False
        self.start_pairing(_TIMEOUT)

    def start_pairing(self, timeout):
        """
        Start the pairing process.

        timeout -- Timeout in seconds at which to quit pairing
        """
        self.pairing = True
        for dev in HapClient.discover(timeout=min(timeout, _TIMEOUT)):
            if not self.pairing:
                break

            _id = 'homekit-' + dev['id']
            if _id not in self.devices:
                try:
                    if dev['ci'] in ['Outlet', 'Switch']:
                        device = HomeKitPlug(self, _id, dev)
                    elif dev['ci'] == 'Lightbulb':
                        device = HomeKitBulb(self, _id, dev)
                    elif dev['ci'] == 'Bridge':
                        device = HomeKitBridge(self, _id, dev)

                        # Don't call handle_device_added(), as we don't want
                        # the bridge itself reported to the gateway.
                        self.devices[device.id] = device
                        continue
                    else:
                        continue
                except PairingError as e:
                    print('Failed to pair with device {}: {}'
                          .format(dev['id'], e))

                    device = HomeKitUnpairedDevice(self, _id, dev)
                except HapError as e:
                    print('Error communicating with device {}: {}'
                          .format(dev['id'], e))
                    continue
                except DatabaseError as e:
                    print('Database error: {}'.format(e))
                    raise

                self.handle_device_added(device)
            elif isinstance(self.devices[_id], HomeKitBridge):
                self.devices[_id].find_new_devices()

    def cancel_pairing(self):
        """Cancel the pairing process."""
        self.pairing = False

    def set_pin(self, device_id, pin):
        """
        Set the PIN for the given device.

        device_id -- ID of device
        pin -- PIN to set
        """
        device = self.get_device(device_id)
        if device:
            if isinstance(device, HomeKitUnpairedDevice):
                try:
                    device.pair(pin)
                    dev = device.dev

                    if dev['ci'] in ['Outlet', 'Switch']:
                        device = HomeKitPlug(self, device_id, dev)
                    elif dev['ci'] == 'Lightbulb':
                        device = HomeKitBulb(self, device_id, dev)
                    elif dev['ci'] == 'Bridge':
                        device = HomeKitBridge(self, device_id, dev)

                    # Replace the unpaired device with the proper device type.
                    self.devices[device_id] = device
                except (DatabaseError, PairingError) as e:
                    print('Failed to pair with device {}: {}'
                          .format(device_id, e))

                    # set_pin() should never be called when SetPinError is not
                    # present, so this is safe.
                    raise SetPinError(str(e))
            else:
                raise SetPinError('Device already paired')

    def remove_thing(self, device_id):
        """
        Unpair a device with the adapter.

        device_id -- ID of device to unpair
        """
        device = self.get_device(device_id)
        if device and not isinstance(device, HomeKitUnpairedDevice):
            if device.bridge is None:
                if device.client.unpair():
                    database = HomeKitDatabase(self.package_name)
                    database.open()
                    database.remove_pairing_data(device.dev['id'])
                    database.close()
            else:
                device.bridge.remove_device(device_id)

            self.handle_device_removed(device)

    def unpair_all(self):
        """Unpair all devices."""
        database = HomeKitDatabase(self.package_name)
        database.open()

        for dev in self.devices.values():
            if isinstance(dev, HomeKitUnpairedDevice) or \
                    dev.bridge is not None:
                continue

            if dev.client.unpair():
                database.remove_pairing_data(dev.dev['id'])

        database.close()

File: lib/homekit-adapter.js
/**
 * HomeKit adapter for Mozilla WebThings Gateway.
 */
'use strict';

const {Adapter} = require('gateway-addon');
const {BLEDiscovery, IPDiscovery} = require('hap-controller');
const HomeKitDatabase = require('./homekit-database');
const HomeKitDevice = require('./homekit-device');
const noble = require('noble');

/**
 * Adapter for HomeKit devices.
 */
class HomeKitAdapter extends Adapter {
  /**
   * Initialize the object.
   *
   * @param {Object} addonManager - AddonManagerProxy object
   * @param {Object} manifest - Package manifest
   */
  constructor(addonManager, manifest) {
    super(addonManager, manifest.name, manifest.name);
    addonManager.addAdapter(this);

    this.currentBLEOperation = Promise.resolve();
    this.queuedOperations = 0;
    this.knownDevices = new Set();

    this.startIPDiscovery();
    this.startBLEDiscovery();

    noble.on('warning', (e) => console.warn('noble warning:', e));
  }

  /**
   * Queue a BLE operation. noble apparently doesn't like operating on more
   * than one device at a time, so we have to queue them up.
   *
   * @param {function} op - Function to add to the queue
   * @returns {Promise} Promise which resolves when the function is called.
   */
  queueBLEOperation(op) {
    this.queuedOperations++;

    const ret = new Promise((resolve, reject) => {
      this.currentBLEOperation.then(() => {
        if (this.bleDiscovery) {
          this.bleDiscovery.stop();
        }

        op().then(() => {
          if (--this.queuedOperations === 0) {
            if (this.bleDiscovery) {
              this.bleDiscovery.start(true);
            }
          }

          resolve.call(arguments);
        }, () => {
          if (--this.queuedOperations === 0) {
            if (this.bleDiscovery) {
              this.bleDiscovery.start(true);
            }
          }

          reject.call(arguments);
        });
      });
    });

    this.currentBLEOperation = ret.catch(() => {});
    return ret;
  }

  /**
   * Set the PIN for the given device.
   *
   * @param {string} deviceId - ID of device
   * @param {string} pin - PIN to set
   * @returns {Promise} Promise which resolves when the PIN has been set.
   */
  setPin(deviceId, pin) {
    const device = this.getDevice(deviceId);
    if (!device) {
      return Promise.reject('Device not found');
    }

    if (device.paired) {
      return Promise.reject('Device already paired');
    }

    if (/^\d{8}$/.test(pin)) {
      pin = `${pin.substr(0, 3)}-${pin.substr(3, 2)}-${pin.substr(5, 3)}`;
    }

    return device.pair(pin);
  }

  /**
   * Unpair a device with the adapter.
   *
   * @param {Object} device - Device to unpair
   * @returns {Promise} Promise which resolves to the device removed.
   */
  removeThing(device) {
    let promise;

    if (device.paired) {
      if (device.bridge) {
        device.bridge.removeDevice(device);
        promise = Promise.resolve();
      } else {
        const database = new HomeKitDatabase(this.packageName);
        promise = device.unpair().then(() => {
          return database.open();
        }).then(() => {
          return database.removePairingData(device.deviceID);
        }).then(() => {
          database.close();
        });
      }
    } else {
      promise = Promise.resolve();
    }

    return promise.then(() => {
      this.knownDevices.delete(device.deviceID);
      this.handleDeviceRemoved(device);
    });
  }

  /**
   * Add a discovered IP device.
   *
   * @param {Object} service - The mDNS service record
   */
  addIPDevice(service) {
    const id = HomeKitDevice.getIdFromService(service, 'ip');

    if (!this.knownDevices.has(id)) {
      console.debug('Found IP device:', service);
      if (service.pv && service.pv.split('.')[0] !== '1') {
        console.info(`Not adding device. PV=${service.pv}`);
        return;
      }

      this.knownDevices.add(id);

      const device = new HomeKitDevice(this, 'ip', service);
      device.promise.then(() => {
        this.handleDeviceAdded(device);
      });
    }
  }

  /**
   * Add a discovered BLE device.
   *
   * @param {Object} service - The BLE advertisement data
   */
  addBLEDevice(service) {
    const id = HomeKitDevice.getIdFromService(service, 'ble');

    if (!this.knownDevices.has(id)) {
      const partial = {
        CoID: service.CoID,
        TY: service.TY,
        AIL: service.AIL,
        SF: service.SF,
        DeviceID: service.DeviceID,
        ACID: service.ACID,
        GSN: service.GSN,
        CN: service.CN,
        CV: service.CV,
      };

      console.debug('Found BLE device:', partial);
      if (service.CV !== 2) {
        console.info(`Not adding device. CV=${service.CV}`);
        return;
      }

      this.knownDevices.add(id);

      const device = new HomeKitDevice(this, 'ble', service);
      device.promise.then(() => {
        this.handleDeviceAdded(device);
      });
    } else {
      // There is a period of time where the device object is being built but
      // is not yet in the devices map.
      const device = this.devices[`homekit-${id}`];
      if (device && device.gsn !== service.GSN) {
        console.debug(`Update triggered for: ${id}`);
        device.triggerBLEUpdate(service.GSN);
      }
    }
  }

  /**
   * Start searching for IP devices.
   */
  startIPDiscovery() {
    this.ipDiscovery = new IPDiscovery();
    this.ipDiscovery.on('serviceUp', (service) => {
      this.addIPDevice(service);
    });
    this.ipDiscovery.start();
  }

  /**
   * Start searching for BLE devices.
   */
  startBLEDiscovery() {
    this.bleDiscovery = new BLEDiscovery();
    this.bleDiscovery.on('serviceUp', (service) => {
      this.addBLEDevice(service);
    });
    this.bleDiscovery.start(true);
  }

  /**
   * Clean up before shutting down this adapter.
   *
   * @returns {Promise} Promise which resolves when finished unloading.
   */
  unload() {
    if (this.ipDiscovery) {
      this.ipDiscovery.stop();
    }

    if (this.bleDiscovery) {
      this.bleDiscovery.stop();
    }

    return super.unload();
  }
}

module.exports = HomeKitAdapter;
